### PR TITLE
fix a tap gesture exception

### LINKED ISSUE #73667: "Null check operator used on a null value" exception when check _down!.buttons in tap.dart
I got huge non-fatal error from firebase about tap.dart.
I cannot reproduce this issue myself.

Here is call stack
```
Non-fatal Exception: FlutterError
0  ???                            0x0 handlePrimaryPointer + 229 (tap.dart:229)
1  ???                            0x0 handleEvent + 476 (recognizer.dart:476)
2  ???                            0x0 _dispatch + 77 (pointer_router.dart:77)
3  ???                            0x0 <fn> + 122 (pointer_router.dart:122)
4  ???                            0x0 forEach (dart:collection)
5  ???                            0x0 _dispatchEventToRoutes + 120 (pointer_router.dart:120)
6  ???                            0x0 route + 106 (pointer_router.dart:106)
7  ???                            0x0 handleEvent + 358 (binding.dart:358)
8  ???                            0x0 dispatchEvent + 338 (binding.dart:338)
9  ???                            0x0 dispatchEvent + 267 (binding.dart:267)
10 ???                            0x0 _handlePointerEvent + 295 (binding.dart:295)
11 ???                            0x0 _flushPointerEventQueue + 240 (binding.dart:240)
12 ???                            0x0 _handlePointerDataPacket + 213 (binding.dart:213)
```

Form source code, it should be `_down!.buttons` _down is null.
https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/gestures/tap.dart
```
  @override
  void handlePrimaryPointer(PointerEvent event) {
    if (event is PointerUpEvent) {
      _up = event;
      _checkUp();
    } else if (event is PointerCancelEvent) {
      resolve(GestureDisposition.rejected);
      if (_sentTapDown) {
        _checkCancel(event, '');
      }
      _reset();
    } else if (event.buttons != _down!.buttons) {
      resolve(GestureDisposition.rejected);
      stopTrackingPointer(primaryPointer!);
    }
  }
```
I have no idea how to fix it, because no call stack from my code.
Any suggestion?

flutter doctor -v
```
[✓] Flutter (Channel stable, 1.22.5, on macOS 11.0.1 20B50 darwin-x64, locale zh-Hant-TW)
    • Flutter version 1.22.5 at /Users/rdapp/git/flutter
    • Framework revision 7891006299 (4 周前), 2020-12-10 11:54:40 -0800
    • Engine revision ae90085a84
    • Dart version 2.10.4

 
[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
    • Android SDK at /Users/rdapp/Library/Android/sdk
    • Platform android-30, build-tools 29.0.2
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 12.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.3, Build version 12C33
    • CocoaPods version 1.10.0

[!] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)
```

### PR DESCRIPTION
**Pull request by xu-baolin**
## Sample Code
```dart
import 'package:flutter/material.dart';

void main() {
  runApp(BugReport());
}

class BugReport extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Directionality(
      textDirection: TextDirection.ltr,
      child: GestureDetector(
        onTap: () {
          print('outer onTap.');
        },
        child: Column(
          children: [
            GestureDetector(
              onDoubleTap: () {
                print('inner onDoubleTap.');
              },
              child: SizedBox(
                width: 100,
                height: 100,
                child: Container(
                  color: const Color(0xFF00FF00),
                ),
              ),
            ),
            SizedBox(
              width: 100,
              height: 100,
              child: Container(
                color: const Color(0xFF0000FF),
              ),
            ),
          ],
        ),
      ),
    );
  }
}
```
## Reproduce Step
Tap one time on the 1st box then tap the 2ond box as much as you can AND quickly.

## Logs
```dart
======== Exception caught by gesture library =======================================================
The following _CastError was thrown while routing a pointer event:
Null check operator used on a null value

When the exception was thrown, this was the stack: 
#0      BaseTapGestureRecognizer.handlePrimaryPointer (package:flutter/src/gestures/tap.dart:229:38)
#1      PrimaryPointerGestureRecognizer.handleEvent (package:flutter/src/gestures/recognizer.dart:475:9)
#2      PointerRouter._dispatch (package:flutter/src/gestures/pointer_router.dart:93:12)
#3      PointerRouter._dispatchEventToRoutes.<anonymous closure> (package:flutter/src/gestures/pointer_router.dart:138:9)
#4      _LinkedHashMapMixin.forEach (dart:collection-patch/compact_hash.dart:397:8)
...
====================================================================================================
```

## Root Cause
The second tap down event will win the tap gesture before the first one because no one competes with it.
Then the recognizer will call `acceptGesture` -> `_checkUp`,
here `_up` is not null because the first point already up and follow into `_reset` and clear `_down`.

At last, the coming move event processing triggers the exception!

Fixes #73667

### CODE REVIEW

**Comment by xu-baolin** [comment]
@dkwingsmt Hi
In fact, the correct behavior of this scene needs to be confirmed, whether the `onTap` callback should be triggered twice or once (ignore the first tap, although it also won the gesture competition).


My solution is to ignore the first tap gesture.

If the right behavior is that each tap will trigger the callback we should let the tap gesture recognizer cash all of the pointers.

**Comment by dkwingsmt** [comment]
I agree with the direction of the behavior, mostly because I think the double tap should be treated as a single tap followed by another single tap, therefore the first single tap should override the standalone tap gesture.
Though, I wonder if there's a better solution to the fix. The current fix looks like you're "patching" an incorrect state; it'll be better if the incorrectness can be detected and found in the first place (i.e. if `_down` is different from `_up` then reset `_up` and do necessary cleanups before accepting `_down`.)
Also, it seems to me that the test can be done on the recognizer level (i.e. using `testGesture` and put into `double_tap_test.dart`). Can you give it a try?

If not, this PR is looking good anyway. Let me know if you to merge as it is.

File: packages/flutter/test/widgets/gesture_detector_test.dart [hunk 1]
```diff
@@ -439,6 +439,51 @@ void main() {
...
+      await tester.pump(const Duration(milliseconds: 50));
+
+      // Tap and move outside container quickly.
+      final TestGesture gesture = await tester.startGesture(const Offset(400.0, 110.0), pointer: 7);
+      await gesture.moveTo(const Offset(400.0, 110.0));
+      await gesture.up();
+      await tester.pumpAndSettle();
+
+      expect(tapCount, 1);
+    },);
```
Comment (dkwingsmt) [inline packages/flutter/test/widgets/gesture_detector_test.dart:486]:
```suggestion
    });
```

File: packages/flutter/test/widgets/gesture_detector_test.dart [hunk 2]
```diff
@@ -439,6 +439,51 @@ void main() {
       await longPress(kLongPressTimeout + const Duration(seconds: 1)); // To make sure the time for long press has occurred
       expect(longPressUp, 1);
     }, variant: buttonVariant);
+
+    testWidgets('Competition', (WidgetTester tester) async {
```
Comment (dkwingsmt) [inline packages/flutter/test/widgets/gesture_detector_test.dart:443]:
```suggestion
    testWidgets('Unfinished DoubleTap does not prevent competing Tap', (WidgetTester tester) async {
```
Or some better names you can think of.

**Comment by xu-baolin** [comment]
But I still have a little doubt. If there is no second tap down (or after a period of time), then the first tap will win the competition and trigger the `ontap` callback, right? Based on this, why does a quick second tap affect the results of the first arena?

**Review by dkwingsmt — APPROVED** [review]
LGTM, the change looks awesome!

**Comment by dkwingsmt** [comment]
@xu-baolin I understand your concern. Although, it's more like
- The DoubleTap grabbed the ownership for the first tap, and Tap no longer has access to it.
- The DoubleTap then ditched it because it didn't receive the 2nd tap. It can not transfer the ownership to Tap by this time either.

If DoubleTap has a "onFirstTap" callback (and it should), this scenario will trigger 2 taps callbacks.

**Pull request merged**